service/qos/standard_service_level_distributed_data_accessor.cc

```diff
@@ -8,27 +8,92 @@
-_sys_dist_ks(sys_dist_ks) {
+static logging::logger dda_logger("service_level_distributed_data_accessor");
+
+static const sstring SERVICE_LEVEL_UPGRADE_KEY = "service_level_upgrade_status";
+static const sstring SL_UPGRADE_DONE_VALUE = "sl_upgraded_v2";
+
+standard_service_level_distributed_data_accessor::standard_service_level_distributed_data_accessor(db::system_distributed_keyspace &sys_dist_ks, gms::feature_service& feat) 
+: _sys_dist_ks(sys_dist_ks)
+, _feat(feat)
+, _v2_upgraded(false) {}
+
+future<> standard_service_level_distributed_data_accessor::init() {
+    auto upgrade_status = co_await db::system_keyspace::get_scylla_local_param(SERVICE_LEVEL_UPGRADE_KEY);
+    
+    if (upgrade_status && *upgrade_status == SL_UPGRADE_DONE_VALUE) {
+        _v2_upgraded = true;
+    } else {
+        if (_data_accessor_upgrade.available()) {
+            _data_accessor_upgrade = upgrade_loop();
+        }
+    }
+}
+
+future<> standard_service_level_distributed_data_accessor::stop() {
+    if (!_data_accessor_upgrade_aborter.abort_requested()) {
+        _data_accessor_upgrade_aborter.request_abort();
+    }
+
+    return std::exchange(_data_accessor_upgrade, make_ready_future<>()).then_wrapped([] (future<> f) {
+        try {
+            f.get();
+        } catch (const broken_semaphore& ignored) {
+        } catch (const sleep_aborted& ignored) {
+        } catch (const exceptions::unavailable_exception& ignored) {
+        } catch (const exceptions::read_timeout_exception& ignored) {
+        }
+    });
 }
 
 future<qos::service_levels_info> standard_service_level_distributed_data_accessor::get_service_levels() const {
-    return _sys_dist_ks.get_service_levels();
+    return (_v2_upgraded) ? _sys_dist_ks.get_service_levels_v2() : _sys_dist_ks.get_service_levels();
 }
 
 future<qos::service_levels_info> standard_service_level_distributed_data_accessor::get_service_level(sstring service_level_name) const {
-    return _sys_dist_ks.get_service_level(service_level_name);
+    return (_v2_upgraded) ? _sys_dist_ks.get_service_level_v2(service_level_name) : _sys_dist_ks.get_service_level(service_level_name);
 }
 
 future<> standard_service_level_distributed_data_accessor::set_service_level(sstring service_level_name, qos::service_level_options slo) const {
-    return _sys_dist_ks.set_service_level(service_level_name, slo);
+    if (!_v2_upgraded) {
+        co_await _sys_dist_ks.set_service_level(service_level_name, slo);
+    }
+    co_await _sys_dist_ks.set_service_level_v2(service_level_name, slo);
 }
 
 future<> standard_service_level_distributed_data_accessor::drop_service_level(sstring service_level_name) const {
-    return _sys_dist_ks.drop_service_level(service_level_name);
+    if (!_v2_upgraded) {
+        co_await _sys_dist_ks.drop_service_level(service_level_name);
+    }
+    co_await _sys_dist_ks.drop_service_level_v2(service_level_name);
+}
+
+future<> standard_service_level_distributed_data_accessor::upgrade_loop() {
```

Comment:
This is a little bit sub optimal but we can do it as a followup, I would expect this upgrade loop to only start once the feature is enabled. (registering a notification on it).
But right now the current mechanism cripples us anyway (this started() function), I would
expect it to be a future or at least to have a future version so we can just wait on it, of course that then we would need to handle aborts differently.
So it is definitely good enough given the other code we interact with, just something to keep in mind :slightly_smiling_face: